pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/managers/ListenerManager.kt

```diff
@@ -65,49 +65,59 @@ class ListenerManager(val pubnub: PubNub) : MessagesConsumer, StatusConsumer, Ev
-        eventListeners.forEach { it.presence(pubnub, presence) }
+        eventListeners.safeForEach { it.presence(pubnub, presence) }
         val envelope = AnnouncementEnvelope(presence)
-        subscriptionCallbacks.forEach { it.presence(pubnub, envelope) }
-        setCallbacks.forEach { it.presence(pubnub, envelope) }
+        subscriptionCallbacks.safeForEach { it.presence(pubnub, envelope) }
+        setCallbacks.safeForEach { it.presence(pubnub, envelope) }
     }
 
     override fun announce(signal: PNSignalResult) {
-        eventListeners.forEach { it.signal(pubnub, signal) }
+        eventListeners.safeForEach { it.signal(pubnub, signal) }
         val envelope = AnnouncementEnvelope(signal)
-        subscriptionCallbacks.forEach { it.signal(pubnub, envelope) }
-        setCallbacks.forEach { it.signal(pubnub, envelope) }
+        subscriptionCallbacks.safeForEach { it.signal(pubnub, envelope) }
+        setCallbacks.safeForEach { it.signal(pubnub, envelope) }
     }
 
     override fun announce(messageAction: PNMessageActionResult) {
-        eventListeners.forEach { it.messageAction(pubnub, messageAction) }
+        eventListeners.safeForEach { it.messageAction(pubnub, messageAction) }
         val envelope = AnnouncementEnvelope(messageAction)
-        subscriptionCallbacks.forEach { it.messageAction(pubnub, envelope) }
-        setCallbacks.forEach { it.messageAction(pubnub, envelope) }
+        subscriptionCallbacks.safeForEach { it.messageAction(pubnub, envelope) }
+        setCallbacks.safeForEach { it.messageAction(pubnub, envelope) }
     }
 
     override fun announce(pnObjectEventResult: PNObjectEventResult) {
-        eventListeners.forEach { it.objects(pubnub, pnObjectEventResult) }
+        eventListeners.safeForEach { it.objects(pubnub, pnObjectEventResult) }
         val envelope = AnnouncementEnvelope(pnObjectEventResult)
-        subscriptionCallbacks.forEach { it.objects(pubnub, envelope) }
-        setCallbacks.forEach { it.objects(pubnub, envelope) }
+        subscriptionCallbacks.safeForEach { it.objects(pubnub, envelope) }
+        setCallbacks.safeForEach { it.objects(pubnub, envelope) }
     }
 
     override fun announce(pnFileEventResult: PNFileEventResult) {
-        eventListeners.forEach { it.file(pubnub, pnFileEventResult) }
+        eventListeners.safeForEach { it.file(pubnub, pnFileEventResult) }
         val envelope = AnnouncementEnvelope(pnFileEventResult)
-        subscriptionCallbacks.forEach { it.file(pubnub, envelope) }
-        setCallbacks.forEach { it.file(pubnub, envelope) }
+        subscriptionCallbacks.safeForEach { it.file(pubnub, envelope) }
+        setCallbacks.safeForEach { it.file(pubnub, envelope) }
+    }
+
+    private inline fun <T> Iterable<T>.safeForEach(action: (T) -> Unit) {
+        for (element in this) {
+            try {
+                action(element)
+            } catch (e: Throwable) {
+                e.printStackTrace()
```

Comment:
What about using logging instead of printStackTrace()?

Reply:
done